File: .cursor/skills/sleuths/sleuth/src/checkpoint.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct SegmentKey {
    pub transcript_id: String,
    pub relative_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Checkpoint {
    #[serde(default)]
    pub processed: Vec<ProcessedSegment>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessedSegment {
    pub transcript_id: String,
    pub relative_path: String,
    pub line_count: u64,
}

impl Checkpoint {
    pub fn load(path: &Path) -> Result<Self> {
        if !path.exists() {
            return Ok(Self::default());
        }
        let raw = std::fs::read_to_string(path)
            .with_context(|| format!("read checkpoint {}", path.display()))?;
        Ok(serde_yaml::from_str(&raw)?)
    }

    pub fn save(&self, path: &Path) -> Result<()> {
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let raw = serde_yaml::to_string(self)?;
        std::fs::write(path, raw)?;
        Ok(())
    }

    pub fn line_count_map(&self) -> HashMap<SegmentKey, u64> {
        self.processed
            .iter()
            .map(|p| {
                (
                    SegmentKey {
                        transcript_id: p.transcript_id.clone(),
                        relative_path: p.relative_path.clone(),
                    },
                    p.line_count,
                )
            })
            .collect()
    }

    pub fn upsert_segment(&mut self, key: SegmentKey, line_count: u64) {
        if let Some(existing) = self.processed.iter_mut().find(|p| {
            p.transcript_id == key.transcript_id && p.relative_path == key.relative_path
        }) {
            existing.line_count = line_count;
        } else {
            self.processed.push(ProcessedSegment {
                transcript_id: key.transcript_id,
                relative_path: key.relative_path,
                line_count,
            });
        }
    }
}

pub fn checkpoint_path(sleuth_dir: &Path) -> PathBuf {
    sleuth_dir.join("checkpoint.yaml")
}

File: .cursor/skills/sleuths/sleuth/checkpoint.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

import yaml


@dataclass(frozen=True)
class SegmentKey:
    transcript_id: str
    relative_path: str


@dataclass
class ProcessedSegment:
    transcript_id: str
    relative_path: str
    line_count: int


@dataclass
class Checkpoint:
    processed: list[ProcessedSegment]

    @classmethod
    def load(cls, path: Path) -> Checkpoint:
        if not path.exists():
            return cls(processed=[])
        with path.open(encoding="utf-8") as f:
            raw = yaml.safe_load(f) or {}
        processed = [
            ProcessedSegment(
                transcript_id=str(p["transcript_id"]),
                relative_path=str(p["relative_path"]),
                line_count=int(p["line_count"]),
            )
            for p in raw.get("processed") or []
        ]
        return cls(processed=processed)

    def save(self, path: Path) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        data = {
            "processed": [
                {
                    "transcript_id": p.transcript_id,
                    "relative_path": p.relative_path,
                    "line_count": p.line_count,
                }
                for p in self.processed
            ]
        }
        with path.open("w", encoding="utf-8") as f:
            yaml.dump(data, f, default_flow_style=False, sort_keys=False)

    def line_count_map(self) -> dict[SegmentKey, int]:
        return {
            SegmentKey(p.transcript_id, p.relative_path): p.line_count for p in self.processed
        }

    def upsert_segment(self, key: SegmentKey, line_count: int) -> None:
        for existing in self.processed:
            if (
                existing.transcript_id == key.transcript_id
                and existing.relative_path == key.relative_path
            ):
                existing.line_count = line_count
                return
        self.processed.append(
            ProcessedSegment(
                transcript_id=key.transcript_id,
                relative_path=key.relative_path,
                line_count=line_count,
            )
        )


def checkpoint_path(sleuth_dir: Path) -> Path:
    return sleuth_dir / "checkpoint.yaml"
